riscv/src/continuations.rs

```diff
@@ -155,45 +156,127 @@ fn sanity_check(main_machine: &Machine, field: KnownField) {
         _ => None,
-    }) else {
+    })
+}
+
+fn expr_to_u32(expr: &Expression) -> Option<u32> {
+    if let Expression::Number(_, Number { value, type_: None }) = expr {
+        value.try_into().ok()
+    } else {
+        None
+    }
+}
+
+pub fn load_initial_memory(program: &AnalysisASMFile, prover_data: &[Vec<u8>]) -> MemoryState {
+    const PAGE_SIZE_BYTES: u32 = bootloader::PAGE_SIZE_BYTES as u32;
+
+    let machine = get_main_machine(program);
+
+    // Extract the prover_data bounds from the machine.
+    let prover_data_start = extract_var_from_machine(machine, "prover_data_start")
+        .expect("prover_data_start variable not found in the machine");
+    let prover_data_start =
+        expr_to_u32(prover_data_start).expect("prover_data_start variable is not a u32 number");
+    let prover_data_end = extract_var_from_machine(machine, "prover_data_end")
+        .expect("prover_data_end variable not found in the machine");
+    let prover_data_end =
+        expr_to_u32(prover_data_end).expect("prover_data_end variable is not a u32 number");
+
+    // Sanity check the bounds of prover_data region.
+    // It must be of a power of 2 size, greater than PAGE_SIZE_BYTES, aligned to its size.
+    let prover_data_size = prover_data_end.checked_sub(prover_data_start).unwrap();
+    assert!(prover_data_size.is_power_of_two());
+    assert!(prover_data_size > PAGE_SIZE_BYTES);
+    assert_eq!(prover_data_start % prover_data_size, 0);
+
+    // Extract the initial_memory variable from the machine.
+    let mut initial_memory = if let Some(expr) = extract_var_from_machine(machine, "initial_memory")
+    {
+        let Expression::ArrayLiteral(_, array) = expr else {
+            panic!("initial_memory is not an array literal");
+        };
+
+        array
+            .items
+            .iter()
+            .map(|entry| {
+                let Expression::Tuple(_, tuple) = entry else {
+                    panic!("initial_memory entry is not a tuple");
+                };
+                assert_eq!(tuple.len(), 2);
+
+                let key = expr_to_u32(&tuple[0]).expect("initial_memory entry key is not a u32");
+                let value =
+                    expr_to_u32(&tuple[1]).expect("initial_memory entry value is not a u32");
+
+                (key, value)
+            })
+            .collect()
+    } else {
         log::warn!("No initial_memory variable found in the machine. Assuming zeroed memory.");
-        return MemoryState::default();
+        MemoryState::default()
     };
 
-    let Expression::ArrayLiteral(_, array) = expr else {
-        panic!("initial_memory is not an array literal");
-    };
+    // Fill the first page with random data to be the salt.
+    // TODO: the random value should be the "hash" of the merkle tree leaf itself,
+    // and the preimage of such hash should be unknown. But to implement that it would
+    // require some inteligence on the MemoryState type.
```

Comment:
I guess this might also affect verification when we actually expose the publics in the bootloader?

Reply:
The idea is to better support secrecy of prover data in case it is not public. Different proofs with the same prover data will yield different merkle tree roots.